specs/IFramePermissionRequested.md

```diff
@@ -0,0 +1,434 @@
+# Background
+The WebView2 team has been asked to provide support for handling permission
+requests that come from iframes. These permission requests occur when content
+within the iframe are requesting access to priveleged resources. The permission
+request types that we support are: Microphone, Camera, Geolocation,
+Notifications, Other Sensors, and Clipboard Read.
+
+We currently have a [PermissionRequested](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2permissionrequestedeventargs?view=webview2-1.0.1020.30)
+event on our `CoreWebView2` which is raised for any permission requests
+(either from webview or iframes). However, our customers do not have a way to
+determine whether the request has come from the webview or an iframe and
+handle these cases seperately. As such, we plan to expand our
+`CoreWebView2Frame` API to include the `PermissionRequested` event.
+
+A current limitation of our `CoreWebView2Frame` API is that it only supports top
+level iframes. Any nested iframes will not have a `CoreWebView2Frame`
+associated with them.
+
+In this document we describe the updated API. We'd appreciate your feedback.
+
+# Description
+We propose extending `CoreWebView2Frame` to include the `PermissionRequested`
+event. This event will be raised whenever the iframe corresponding to the
+CoreWebView2Frame or any of its descendant iframes requests permission to
+priveleged resources.
+
+Additionally, we propose extending `CoreWebView2PermissionRequestedEventArgs`
+to include a `Handled` property.
+
+To maintain backwards compatibility, by default we plan to raise
+`PermissionRequested` on both `CoreWebView2Frame` and `CoreWebView2`. The
+`CoreWebView2Frame` event handlers will be invoked first,
+before the `CoreWebView2` event handlers. If `Handled` is set true as part of
+the `CoreWebView2Frame` event handlers, then the `PermissionRequested` event
+will not be raised on the `CoreWebView2`, and its event handlers will not be
+invoked.
```

Comment:
I think this design (raising two events) creates unnecessary complexity, because people will wonder which one they should register for, and you now have to document who gets the event first, what it means to set Handled on the document-level event, what happens if the handlers disagree, etc. And then when someday you decide to raise the PermissionRequested event on the nested requesting frame, we will have to raise *three* events, with even more documentation on the order of the events, what Handled means, and who wins in case of conflict.

Why not just funnel everything through the existing event, but with new properties that tell you the requesting frame?

* New property CoreWebView2PermissionRequestedEventArgs.RequestingTopFrame. It is the top frame making the permission request, or `nullptr` if the request is coming from the main document.

Old apps will ignore the RequestingTopFrame property and behave as before; They receive all permission requests on the main document.

New apps will check the RequestingTopFrame to see which top frame contains the requesting frame.

If someday we decide to tell you the specific requesting frame (even if nested), we can add a second new property `RequestingFrame`.

No need for a `Handled` property, since there is only one event.

Reply:
Following frame navigation precedent and having examined all the other events on the corewebview2 there are no other events that are frame level but also apply to all child frames. So this should be the last of this kind of event.